// @nrwl/tao/src/shared/package-manager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPackageManagerCommand = exports.detectPackageManager = void 0;
const fs_1 = require("fs");
function detectPackageManager() {
    return fs_1.existsSync('yarn.lock')
        ? 'yarn'
        : fs_1.existsSync('pnpm-lock.yaml')
            ? 'pnpm'
            : 'npm';
}
exports.detectPackageManager = detectPackageManager;
function getPackageManagerCommand(packageManager = detectPackageManager()) {
    switch (packageManager) {
        case 'yarn':
            return {
                install: 'yarn',
                add: 'yarn add',
                addDev: 'yarn add -D',
                rm: 'yarn rm',
                exec: 'yarn',
                list: 'yarn list',
            };
        case 'pnpm':
            return {
                install: 'pnpm install',
                add: 'pnpm add',
                addDev: 'pnpm add -D',
                rm: 'pnpm rm',
                exec: 'pnpx',
                list: 'pnpm ls --depth 100',
            };
        case 'npm':
            return {
                install: 'npm install',
                add: 'npm install',
                addDev: 'npm install -D',
                rm: 'npm rm',
                exec: 'npx',
                list: 'npm ls',
            };
    }
}
exports.getPackageManagerCommand = getPackageManagerCommand;
