// @nrwl/tao/src/shared/tree.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.printChanges = exports.flushChanges = exports.FsTree = void 0;
const path = require("path");
const fs_1 = require("fs");
const fs_extra_1 = require("fs-extra");
const logger_1 = require("./logger");
const chalk = require('chalk');
class FsTree {
    constructor(root, isVerbose) {
        this.root = root;
        this.isVerbose = isVerbose;
        this.recordedChanges = {};
    }
    read(filePath) {
        try {
            if (this.recordedChanges[this.rp(filePath)]) {
                return this.recordedChanges[this.rp(filePath)].content;
            }
            else {
                return this.fsReadFile(filePath);
            }
        }
        catch (e) {
            if (this.isVerbose) {
                logger_1.logger.error(e);
            }
            return null;
        }
    }
    write(filePath, content) {
        try {
            this.recordedChanges[this.rp(filePath)] = {
                content: Buffer.from(content),
                isDeleted: false,
            };
        }
        catch (e) {
            if (this.isVerbose) {
                logger_1.logger.error(e);
            }
        }
    }
    overwrite(filePath, content) {
        this.write(filePath, content);
    }
    exists(filePath) {
        try {
            if (this.recordedChanges[this.rp(filePath)]) {
                return !this.recordedChanges[this.rp(filePath)].isDeleted;
            }
            else if (this.filesForDir(this.rp(filePath)).length > 0) {
                return true;
            }
            else {
                return this.fsExists(filePath);
            }
        }
        catch (err) {
            return false;
        }
    }
    delete(filePath) {
        if (this.filesForDir(this.rp(filePath)).length > 0) {
            this.filesForDir(this.rp(filePath)).forEach((f) => (this.recordedChanges[f] = { content: null, isDeleted: true }));
        }
        this.recordedChanges[this.rp(filePath)] = {
            content: null,
            isDeleted: true,
        };
    }
    rename(from, to) {
        const content = this.read(this.rp(from));
        this.recordedChanges[this.rp(from)] = { content: null, isDeleted: true };
        this.recordedChanges[this.rp(to)] = { content: content, isDeleted: false };
    }
    isFile(filePath) {
        try {
            if (this.recordedChanges[this.rp(filePath)]) {
                return !this.recordedChanges[this.rp(filePath)].isDeleted;
            }
            else {
                return this.fsIsFile(filePath);
            }
        }
        catch (err) {
            return false;
        }
    }
    children(dirPath) {
        let res = this.fsReadDir(dirPath);
        res = [...res, ...this.directChildrenOfDir(this.rp(dirPath))];
        return res.filter((q) => {
            const r = this.recordedChanges[path.join(this.rp(dirPath), q)];
            if (r && r.isDeleted)
                return false;
            return true;
        });
    }
    listChanges() {
        const res = [];
        Object.keys(this.recordedChanges).forEach((f) => {
            if (this.recordedChanges[f].isDeleted) {
                if (this.fsExists(f)) {
                    res.push({ path: f, type: 'DELETE', content: null });
                }
            }
            else {
                if (this.fsExists(f)) {
                    res.push({
                        path: f,
                        type: 'UPDATE',
                        content: this.recordedChanges[f].content,
                    });
                }
                else {
                    res.push({
                        path: f,
                        type: 'CREATE',
                        content: this.recordedChanges[f].content,
                    });
                }
            }
        });
        return res;
    }
    fsReadDir(dirPath) {
        if (!this.delegateToFs)
            return [];
        try {
            return fs_1.readdirSync(path.join(this.root, dirPath));
        }
        catch (e) {
            return [];
        }
    }
    fsIsFile(filePath) {
        if (!this.delegateToFs)
            return false;
        const stat = fs_1.statSync(path.join(this.root, filePath));
        return stat.isFile();
    }
    fsReadFile(filePath) {
        if (!this.delegateToFs)
            return null;
        return fs_1.readFileSync(path.join(this.root, filePath));
    }
    fsExists(filePath) {
        if (!this.delegateToFs)
            return false;
        try {
            const stat = fs_1.statSync(path.join(this.root, filePath));
            return stat.isFile() || stat.isDirectory();
        }
        catch (e) {
            return false;
        }
    }
    delegateToFs() {
        return this.root !== null;
    }
    filesForDir(path) {
        return Object.keys(this.recordedChanges).filter((f) => f.startsWith(path + '/') && !this.recordedChanges[f].isDeleted);
    }
    directChildrenOfDir(path) {
        const res = {};
        Object.keys(this.recordedChanges).forEach((f) => {
            if (f.startsWith(path + '/')) {
                const [_, file] = f.split(path + '/');
                res[file.split('/')[0]] = true;
            }
        });
        return Object.keys(res);
    }
    rp(pp) {
        return pp.startsWith('/') ? pp.substring(1) : pp;
    }
}
exports.FsTree = FsTree;
function flushChanges(root, fileChanges) {
    fileChanges.forEach((f) => {
        const fpath = path.join(root, f.path);
        if (f.type === 'CREATE') {
            fs_extra_1.mkdirpSync(path.dirname(fpath));
            fs_1.writeFileSync(fpath, f.content);
        }
        else if (f.type === 'UPDATE') {
            fs_1.writeFileSync(fpath, f.content);
        }
        else if (f.type === 'DELETE') {
            try {
                const stat = fs_1.statSync(fpath);
                if (stat.isDirectory()) {
                    fs_extra_1.rmdirSync(fpath, { recursive: true });
                }
                else {
                    fs_1.unlinkSync(fpath);
                }
            }
            catch (e) { }
        }
    });
}
exports.flushChanges = flushChanges;
function printChanges(fileChanges) {
    fileChanges.forEach((f) => {
        if (f.type === 'CREATE') {
            console.log(`${chalk.green('CREATE')} ${f.path}`);
        }
        else if (f.type === 'UPDATE') {
            console.log(`${chalk.white('UPDATE')} ${f.path}`);
        }
        else if (f.type === 'DELETE') {
            console.log(`${chalk.yellow('DELETE')} ${f.path}`);
        }
    });
}
exports.printChanges = printChanges;
