// @nrwl/workspace/src/schematics/run-commands/run-commands.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.runCommandsGenerator = void 0;
const workspace_1 = require("@nrwl/workspace");
const ngcli_adapter_1 = require("@nrwl/devkit/ngcli-adapter");
function default_1(host, schema) {
    const fn = workspace_1.updateWorkspaceInTree((json) => {
        const project = json.projects[schema.project];
        if (!project) {
            throw new Error(`Invalid project name "${schema.project}"`);
        }
        project.architect = project.architect || {};
        project.architect[schema.name] = {
            builder: '@nrwl/workspace:run-commands',
            outputs: schema.outputs
                ? schema.outputs.split(',').map((s) => s.trim())
                : [],
            options: {
                command: schema.command,
                cwd: schema.cwd,
                envFile: schema.envFile,
            },
        };
        return json;
    });
    fn(host);
}
exports.default = default_1;
exports.runCommandsGenerator = ngcli_adapter_1.wrapAngularDevkitSchematic('@nrwl/workspace', 'run-commands');
