// @nrwl/jest/src/schematics/jest-project/lib/update-workspace.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.updateWorkspace = void 0;
const core_1 = require("@angular-devkit/core");
const devkit_1 = require("@nrwl/devkit");
function updateWorkspace(tree, options) {
    var _a;
    const projectConfig = devkit_1.readProjectConfiguration(tree, options.project);
    projectConfig.targets.test = {
        executor: '@nrwl/jest:jest',
        outputs: [core_1.join(core_1.normalize('coverage'), core_1.normalize(projectConfig.root))],
        options: {
            jestConfig: core_1.join(core_1.normalize(projectConfig.root), 'jest.config.js'),
            passWithNoTests: true,
        },
    };
    const isUsingTSLint = ((_a = projectConfig.targets.lint) === null || _a === void 0 ? void 0 : _a.executor) ===
        '@angular-devkit/build-angular:tslint';
    if (isUsingTSLint) {
        projectConfig.targets.lint.options.tsConfig = [
            ...projectConfig.targets.lint.options.tsConfig,
            core_1.join(core_1.normalize(projectConfig.root), 'tsconfig.spec.json'),
        ];
    }
    devkit_1.updateProjectConfiguration(tree, options.project, projectConfig);
}
exports.updateWorkspace = updateWorkspace;
